### ISSUE TITLE
handle the naturally missing dates

### ISSUE THREAD

**Issue by e-kotov** [rOpenSpain/spanishoddata#90]
V2 data has some missing dates [source](https://www.transportes.gob.es/ministerio/proyectos-singulares/estudios-de-movilidad-con-big-data/opendata-movilidad):
_"Data publication will resume in November, continuing the series. Days not available due to incidence in Orange: October 26,27,30,31, November 1, 2 and 3, 2023 and April 4,18,19, 2024."_ (Though that's not entirely accurate, I checked, and April 4 is actually available.)

Therefore:

```r
data_missing <- spod_get(
 type = "number_of_trips",
 zones = "distr",
 dates = c(start = "2023-11-01", end = "2023-11-04")
)
```
Gives:

```
Error in `spod_infer_data_v_from_dates()`:
! Some dates do not match the available data. 
The valid dates range for v1 is: 2020-02-14 to 2021-05-09 and for v2 is: 2022-01-01 to 2024-06-30.
Missing dates: 2023-11-01, 2023-11-02, 2023-11-03
```

@Robinlovelace, @eugenividal , I need your opinion here.

What do we do with that..?

Options (note, all of these only relate to the "official" missing dates, not just any dates that are out of range or intersect between v1 and v2 data. These out of range dates should cause the error above as intended):

### 1. Ignore Missing Dates with a Warning

Keep a list of these "officially" missing dates and ignore them in any period request that involves them. Let the function just return whichever dates are available and skip the ones that are not. Add a warning that will be force-printed (ignoring the `quiet=TRUE` argument if set) when that happens, informing the user that these dates are impossible to fetch and link to the Ministry's website.

- Pros:

  - Simplifies the user experience by automatically handling missing dates.
 
  - Provides a warning to inform the user about the missing dates.
  
- Cons:

  - Users may overlook warnings, especially in non-interactive sessions.

  - Silent data omission could lead to inaccurate analyses if users are unaware (e.g. these missing days will absolutely affect the results of a simple "total trips per month" summary, which is bad...)

### 2. Provide an Option to Ignore Missing Dates

Keep the current behavior but introduce an option (function argument) to ignore "officially" missing dates. This way, we urge the user to notice the problem and be aware of it, because without this parameter set explicitly, they will get error. I think this may be more efficient in preventing users from making mistakes. In this case, I would just update the current error message we have to inform about why dates are missing and how to acknowledge it and proceed with the requested date interval as is.

- Pros:

  - Forces users to consciously decide how skip missing dates.
  
  - Encourages users to be aware of data completeness.

- Cons:

  - Adds complexity to the function interface with additional arguments (but not too much).

  - Users might still ignore the issue by setting the option without understanding the implications (but we did the best we could to inform them by throwing an error).

### 3. Provide a Function to Adjust Date Intervals

Give users a function that can convert their desired date intervals into valid ones that don't contain "officially" missing dates and will pass the data availability checks built into all other functions in the package. I don't like this option, as it encourages the user to just pass all dates intervals through this magic function. Though it is not that much different from option 2.

- Pros:

  - Helps users generate valid date ranges programmatically.

  - Ensures that data retrieval functions operate smoothly.

- Cons:

  - Hides the issue of missing data from users if they make a habit to just use this function with any date interval.
  
  - Users may remain unaware of gaps in their data, leading to flawed analyses, but in fact same as option 2.

### 4. Interactive Prompt for Missing Dates

If the session is interactive, prompt the user to decide how to handle missing dates when they are detected.

- Pros:

  - Ensures users are immediately aware of the issue.

  - Provides an opportunity to educate users about data gaps.

- Cons:

  - Interrupts automated scripts or batch processing.

  - Not suitable for non-interactive sessions.

# Summary

I think none of these options are perfect, but I like option 2. At least, I think, with any of the options we are still doing a little bit better than just having this information in the codebook.

@Robinlovelace , @eugenividal , what do you think?

**Comment by Robinlovelace** [comment]
2) would get my vote.

**Comment by Robinlovelace** [comment]
Reasoning, I think the pros are big and cons relatively small:


> Pros:
> 
>     * Forces users to consciously decide how skip missing dates.

This is a big one. Adding silent things to make lives easier short term can lead to long-term issues, especially in relation to statistical analysis. Failing early is good I think.
> 
>     * Encourages users to be aware of data completeness.

Same



> Cons:
> 
>     * Adds complexity to the function interface with additional arguments (but not too much).

It's just a single argument, and all solutions would add some complexity (this one relatively little I believe).

> 
>     * Users might still ignore the issue by setting the option without understanding the implications (but we did the best we could to inform them by throwing an error).

Is that really a con? They're less likely to ignore it if there's a big Error message than if not.

**Comment by eugenividal** [comment]
@e-kotov I took a quick look at this and I agree with you guys, I think the preference should be to prevent users from making mistakes.

**Comment by e-kotov** [comment]
> This is a big one. Adding silent things to make lives easier short term can lead to long-term issues, especially in relation to statistical analysis. Failing early is good I think.

Yes, that was my thinking too. I am ready to sacrifice some convenience and smoothness of using the package to prevent some bitter mistakes.
